sqld/src/namespace.rs

```diff
@@ -0,0 +1,304 @@
+pub trait NamespaceFactory: Sync + Send + 'static {
+    type Database: Database;
+    type Meta: Send + Sync + 'static;
+
+    async fn create(&self, name: Bytes) -> anyhow::Result<Namespace<Self::Database, Self::Meta>>;
+}
+
+pub struct PrimaryNamespaceFactory {
+    config: PrimaryNamespaceConfig,
+}
+
+impl PrimaryNamespaceFactory {
+    pub fn new(config: PrimaryNamespaceConfig) -> Self {
+        Self { config }
+    }
+}
+
+#[async_trait::async_trait]
+impl NamespaceFactory for PrimaryNamespaceFactory {
+    type Database = TrackedDb<LibSqlDb>;
+    type Meta = PrimaryMeta;
+
+    async fn create(&self, name: Bytes) -> anyhow::Result<Namespace<Self::Database, Self::Meta>> {
+        Namespace::new_primary(&self.config, name).await
+    }
+}
+
+pub struct ReplicaNamespaceFactory {
+    config: ReplicaNamespaceConfig,
+}
+
+impl ReplicaNamespaceFactory {
+    pub fn new(config: ReplicaNamespaceConfig) -> Self {
+        Self { config }
+    }
+}
+
+#[async_trait::async_trait]
+impl NamespaceFactory for ReplicaNamespaceFactory {
+    type Database = TrackedDb<WriteProxyDatabase>;
+    type Meta = ();
+
+    async fn create(&self, name: Bytes) -> anyhow::Result<Namespace<Self::Database, Self::Meta>> {
+        Namespace::new_replica(&self.config, name).await
+    }
+}
+
+pub struct Namespaces<F: NamespaceFactory> {
+    #[allow(clippy::type_complexity)]
+    inner: RwLock<HashMap<Bytes, Namespace<F::Database, F::Meta>>>,
+    factory: F,
+}
+
+impl<F: NamespaceFactory> Namespaces<F> {
+    pub fn new(factory: F) -> Self {
+        Self {
+            inner: Default::default(),
+            factory,
+        }
+    }
+
+    pub async fn reset(&self, namespace: Bytes) -> anyhow::Result<()>
+    where
+        Namespace<F::Database, F::Meta>: NamespaceCommon,
+    {
+        let mut lock = self.inner.write().await;
+        if let Some(ns) = lock.remove(&namespace) {
+            // FIXME: when destroying, we are waiting for all the tasks associated with the
+            // allocation to finnish, which create a lot of contention on the lock. Need to use a
+            // conccurent hashmap to deal with this issue.
+            ns.destroy().await?;
+            // re-create the namespace
+            let ns = self.factory.create(namespace.clone()).await?;
+            lock.insert(namespace, ns);
+        }
+
+        Ok(())
+    }
+
+    pub async fn with<Fun, R>(&self, namespace: Bytes, f: Fun) -> anyhow::Result<R>
+    where
+        Fun: FnOnce(&Namespace<F::Database, F::Meta>) -> R,
+    {
+        let lock = self.inner.upgradable_read().await;
+        if let Some(ns) = lock.get(&namespace) {
+            Ok(f(ns))
+        } else {
+            let mut lock = RwLockUpgradableReadGuard::upgrade(lock).await;
+            let ns = self.factory.create(namespace.clone()).await?;
+            let ret = f(&ns);
+            lock.insert(namespace, ns);
+            Ok(ret)
+        }
+    }
+}
+
+#[allow(dead_code)]
+pub struct Namespace<D, T> {
```

Comment:
This name as well doesn't make sense to me either would be good to have a doc comment or rename it to something more digestible.